pkg/scheduler/controller/scheduler_control.go

```diff
@@ -0,0 +1,375 @@
+	if shoot.Spec.Cloud.Seed != nil {
+		if err := validateSeed(shoot, c.seedLister); err != nil {
+			message := fmt.Sprintf("Shoot '%s': Validation of seed '%s' failed: %+v", shoot.Name, *shoot.Spec.Cloud.Seed, err)
+			c.updateShootStatusWithCondition(ctx, shoot, message, conditionSeedAvailable)
+			return err
+		}
+		return nil
+	}
+
+	schedulerLogger.Infof("[SCHEDULING SHOOT] using %s strategy", c.config.Strategy)
+
+	// If no Seed is referenced, we try to determine an adequate one.
+	seed, err := determineSeed(shoot, c.seedLister, c.shootLister, c.config.Strategy)
+	if err != nil {
+		c.updateShootStatusToUnschedulable(ctx, shoot, err, conditionSeedAvailable)
+		return err
+	}
+
+	updateBinding := func(ctx context.Context, binding *gardenv1beta1.ShootBinding) error {
+		return c.k8sGardenClient.Garden().GardenV1beta1().Shoots(shoot.Namespace).Bind(binding)
+	}
+
+	if err := UpdateShootToBeScheduledOntoSeed(ctx, shoot, seed, updateBinding); err != nil {
+		c.updateShootStatusToUnschedulable(ctx, shoot, err, conditionSeedAvailable)
+		return err
+	}
+
+	schedulerLogger.Infof("Shoot '%s' (Cloud Profile '%s', Region '%s') successfully scheduled to seed '%s' using SeedDeterminationStrategy '%s'", shoot.Name, shoot.Spec.Cloud.Profile, shoot.Spec.Cloud.Region, seed.Name, c.config.Strategy)
+	return nil
+}
+
+func validateSeed(shoot *gardenv1beta1.Shoot, seedLister gardenlisters.SeedLister) error {
+	seed, err := seedLister.Get(*shoot.Spec.Cloud.Seed)
+	if err != nil {
+		return err
+	}
+	if shoot.Namespace != common.GardenNamespace && seed.Spec.Protected != nil && *seed.Spec.Protected {
+		return fmt.Errorf("forbidden to use a protected seed")
+	}
+
+	if seed.DeletionTimestamp != nil {
+		return fmt.Errorf("forbidden to use a seed marked to be deleted")
+	}
+	if hasDisjointedNetworks, allErrs := validateDisjointedNetworks(seed, shoot); !hasDisjointedNetworks {
+		return allErrs.ToAggregate()
+	}
+	return nil
+}
+
+// determineSeed returns an appropriate Seed cluster (or nil).
+func determineSeed(shoot *gardenv1beta1.Shoot, seedLister gardenlisters.SeedLister, shootLister gardenlisters.ShootLister, strategy config.CandidateDeterminationStrategy) (*gardenv1beta1.Seed, error) {
+	seedList, err := seedLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	shootList, err := shootLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	return determineBestSeedCandidate(shoot, shootList, seedList, strategy)
+}
+
+func determineBestSeedCandidate(shoot *gardenv1beta1.Shoot, shootList []*gardenv1beta1.Shoot, seedList []*gardenv1beta1.Seed, strategy config.CandidateDeterminationStrategy) (*gardenv1beta1.Seed, error) {
+	// Map seeds to number of managed shoots.
+	var (
+		seedUsage  = generateSeedUsageMap(shootList)
+		candidates []*gardenv1beta1.Seed
+	)
+
+	switch strategy {
+	case config.SameRegion:
+		candidates = determineCandidatesWithSameRegionStrategy(seedList, shoot, candidates)
```

Comment:
Instead of having this `switch` here, can't the `strategy` directly implement filtering the potential candidates?

Reply:
Currently i would be fine with the solution but it might make sense when we add more strategies? what do you think @rfranzke ?